### PR TITLE
Use a BOP for Hash#default

### PR DESCRIPTION
**Pull request by jhawthorn**
On a hash miss we need to call default if it is redefined in order to return the default value to be used. Previously we checked this with rb_method_basic_definition_p, which avoids the method call but requires a method lookup.

This commit replaces the previous check with BASIC_OP_UNREDEFINED_P and a new BOP_DEFAULT. We still need to fall back to rb_method_basic_definition_p when called on a subclass of hash.

    |                |compare-ruby|built-ruby|
    |:---------------|-----------:|---------:|
    |hash_aref_miss  |       2.692|     3.531|
    |                |           -|     1.31x|

On railsbench, before this change I measured how many times we call `rb_method_basic_definition_p` and with what IDs. For 5 iterations:

**before:** 5086076
**after:** 716854

So this is optimizing ~95% of the checks. Rails does include some subclasses of Hash (ex. HashWithIndifferentAccess), which is why it isn't 100%. `default` was by far the ID most passed to rb_method_basic_definition_p, second was `to_proc` with 273898 calls.

### CODE REVIEW

**Comment by k0kubun** [comment]
> This does seem to improve Railsbench a little bit

nitpick: It seems more easier to understand the difference this PR is making if you compare before/after between interpreters and YJITs, not interp/yjit. You could do it like:

```
./run_benchmarks.rb railsbench -e "/path/to/old-ruby" -e "/path/to/new-ruby"
./run_benchmarks.rb railsbench -e "/path/to/old-ruby --yjit" -e "/path/to/new-ruby --yjit"
```

If you intentionally do it that way to avoid having multiple build directories, it's okay too :) But I just wanted to make sure you're aware.

**Comment by jhawthorn** [comment]
I should do that 😅. When I'm testing just for myself I tend to run the comparison Ruby directly from the source tree, but that has also caused me to make mistakes.

File: hash.c [hunk 1]
```diff
@@ -2070,12 +2073,25 @@ call_default_proc(VALUE proc, VALUE hash, VALUE key)
     return rb_proc_call_with_block(proc, 2, args, Qnil);
 }
 
+bool
```
Comment (XrXr) [inline hash.c:2076]:
```suggestion
static bool
```

File: hash.c [hunk 2]
```diff
@@ -2070,12 +2073,25 @@ call_default_proc(VALUE proc, VALUE hash, VALUE key)
...
+rb_hash_default_unredefined(VALUE hash)
+{
+    VALUE klass = RBASIC_CLASS(hash);
+    if (LIKELY(klass == rb_cHash)) {
+        return !!BASIC_OP_UNREDEFINED_P(BOP_DEFAULT, HASH_REDEFINED_OP_FLAG);
+    } else {
```
Comment (XrXr) [inline hash.c:2082]:
```suggestion
    }
    else {
```

**Review by XrXr — COMMENTED** [review]
Nice

**Pull request merged**